### PR TITLE
Start agent

### PR DESCRIPTION
**Pull request by EderSantana**
I did this before our discussion on about main loop (ie. fit). I wonder how much we should define in the agent `__init__` and how much we should leave for `fit`...

### CODE REVIEW

File: keras_dqn/agent.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+from keras.utils.np_utils import to_categorical
+
+
+class Agent(object):
+    def __init__(self, model, target_model=None, target_swap=None):
```
Comment (ntasfi) [inline keras_dqn/agent.py]: does the agent accept all the training parameters? Or will it be external to the agent? eg. if we do annealing of the learning rate etc.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    package_dir={'keras_dqn':
+                 'keras_dqn'},
+    include_package_data=True,
+    install_requires=requirements,
+    license="MIT?",
+    zip_safe=False,
+    keywords='x',
+    classifiers=[
+        'Development Status :: 2 - Pre-Alpha',
+        'Intended Audience :: Witches',
```
Comment (ntasfi) [inline setup.py:45]: haha :100:

File: keras_dqn/agent.py (hunk 1 above)
Comment (ntasfi) [inline keras_dqn/agent.py]: also might be useful to accept an RNG value here too

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    author="ntasfi and edersantana",
+    url='https://github.com/ntasfi/keras_dqn',
+    packages=[
+        'keras_dqn',
+    ],
+    package_dir={'keras_dqn':
+                 'keras_dqn'},
+    include_package_data=True,
+    install_requires=requirements,
+    license="MIT?",
```
Comment (ntasfi) [inline setup.py]: MIT is fine with me

File: keras_dqn/agent.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        # gradient step
+        if accum_mode == "mean":
+            self.model.train_on_batch(state, target, sample_weight=None)
+
+    def update_target(self):
+        if self.target_swap:
+            weights = self.model.get_weights()
+            self.target_model.set_weights(weights)
+
+    def fit(env, batch_size, epochs):
```
Comment (ntasfi) [inline keras_dqn/agent.py:85]: will we have a `play` or `watch` function? where you can pass a game network weights file and have it play?

File: keras_dqn/agent.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        """Learn to play
+        Extra desired args:
```
Comment (ntasfi) [inline keras_dqn/agent.py:87]: should some of these be passed to the `__init__` function? That way we can pickle the agent+model easily?

File: keras_dqn/agent.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        return self.model.output_shape[-1]
+
+    def train_on_batch(self, state, action, reward, next_state, terminal, gamma=.99, accum_mode="mean"):
+        q = self.model.predict(state)
+        next_max_q = np.argmax(self.target_model.predict(next_state), axis=-1)
+        idx = to_categorical(action, self.num_actions)  # transform actions to one-hot indices
+
+        # target = reward + gamma * max Q(s', a')
+        target = np.copy(q)  # we will use indexing next, avoid aliasing
+        # @ntasfi, regardless of what is target (1-terminal) zeros it out that is why I thought we don't have to store terminal states
```
Comment (ntasfi) [inline keras_dqn/agent.py]: I was thinking this over too. It might be best to leave it in, as it currently doesn't hurt and it might have use for people if they were tweaking or extending this out

File: keras_dqn/agent.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        else:
+            self.target_model = target_model
+            self.target_swap = True
+            print("Using target model update.")
+
+    @property
+    def num_actions(self):
+        return self.model.output_shape[-1]
+
+    def train_on_batch(self, state, action, reward, next_state, terminal, gamma=.99, accum_mode="mean"):
```
Comment (ntasfi) [inline keras_dqn/agent.py]: can we include the shape expected here? I can make the Memory module obey what we decide on here

File: keras_dqn/agent.py (hunk 1 above)
Comment (ntasfi) [inline keras_dqn/agent.py]: do we want the agent to handle the model and target_model? Or should all those functions just be pushed into the network file?

File: keras_dqn/agent.py (hunk 4 above)
Comment (EderSantana) [inline keras_dqn/agent.py:85]: yeah, evaluate, play or watch. what do you prefer?

File: keras_dqn/agent.py (hunk 5 above)
Comment (EderSantana) [inline keras_dqn/agent.py:87]: I think so... that is what I'm doing with X. It feels much better to have everything in the `__init__`

File: keras_dqn/agent.py (hunk 6 above)
Comment (EderSantana) [inline keras_dqn/agent.py]: true!

File: keras_dqn/agent.py (hunk 7 above)
Comment (EderSantana) [inline keras_dqn/agent.py]: ok!

File: setup.py (hunk 2 above)
Comment (EderSantana) [inline setup.py:45]: don't let me forget to take this out 💃

File: keras_dqn/agent.py (hunk 1 above)
Comment (EderSantana) [inline keras_dqn/agent.py]: network will be just keras models... to avoid creating an extra class, let us put everything into the agent. What do you think?

File: keras_dqn/agent.py (hunk 4 above)
Comment (ntasfi) [inline keras_dqn/agent.py:85]: play or evaluate! you pick

**Comment by EderSantana** [comment]
how are we handling data renormalization and reward clipping? should they be parameters on the input of agent or callbacks?

**Comment by ntasfi** [comment]
we could perhaps do the data renormalization and reward clipping inside of the memory? It really depends on what is orchestrating all the pieces though!

File: keras_dqn/agent.py (hunk 5 above)
Comment (ntasfi) [inline keras_dqn/agent.py:87]: looks good!

**Comment by EderSantana** [comment]
yeah, if we get the agent to orchestrate everything its like a centralized set of parameters. I was spreading parameters through all the classes and that was hard to track

**Comment by ntasfi** [comment]
Awesome, so this looks like a good starting point! Do you want to work on the q_network or should I?

File: keras_dqn/agent.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        if target_model is None:
+            self.target_model = model
+        else:
+            self.target_model = target_model
+            self.target_swap = True  # external target flag
+            print("Using target model update.")
+        self.update_model_every = update_model_every
+        self.update_target_every = update_target_every
+
+        # memory info
```
Comment (ntasfi) [inline keras_dqn/agent.py]: Should have max_frames here too?

File: keras_dqn/agent.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        elif train is False and rnd < self.test_eps:
+            return rnd_action
+        else:
+            # this code is meant to work only if doing a single action!
+            q = self.model.predict(observation)
+            # random tie breaker
+            max_inds = np.argwhere(q == np.max(q))
+            return np.random.choice(max_inds.reshape(-1,))
+
+    def train_on_batch(self, state, action, reward, next_state, terminal, gamma=.99, accum_mode="mean"):
```
Comment (ntasfi) [inline keras_dqn/agent.py]: Will we have the agent create and handle the memory internally or externally? If its internally we should need to pass state, action, etc in, or we can and just make this a private function?

Comment (EderSantana) [inline keras_dqn/agent.py]: yeah I should have used an `_` in there
